geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,236 @@
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: N_STEPS.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 2, M.dim]
+            Point on the tangent bundle TM.
+        """
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+        tngs = gs.reshape(tangent_vec, bs_pts.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        eps = 1 / n_steps
+
+        v0, w0 = gs.take(tngs, 0, axis=1), gs.take(tngs, 1, axis=1)
+        p0, u0 = gs.take(bs_pts, 0, axis=1), gs.take(bs_pts, 1, axis=1)
+        for _ in range(n_steps):
+            p = metric.exp(eps * v0, p0)
+            u = par_trans(u0 + eps * w0, p0, None, p)
+            v = par_trans(v0 - eps * (metric.curvature(u0, w0, v0, p0)), p0, None, p)
+            w = par_trans(w0, p0, None, p)
+            p0, u0 = p, u
+            v0, w0 = v, w
+
+        return gs.reshape(gs.stack([p, u], axis=1), base_point.shape)
+
+    def log(self, point, base_point, n_steps=N_STEPS, **kwargs):
+        """Compute the Riemannian logarithm of a point.
+
+        Logarithmic map at base_point of point computed by
+        iteratively relaxing a discretized geodesic between base_point and point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: N_STEPS.
+
+        Returns
+        -------
+        log : array-like, shape=[..., 2, M.dim]
+            Tangent vector at the base point equal to the Riemannian logarithm
+            of point at the base point.
+        """
+        pts = gs.reshape(point, (-1, 2) + self.metric.shape)
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+
+        @delayed
+        def _log(pt, bs_pt):
+            """Calculate the logarithm."""
+            pu = self.geodesic_discrete(bs_pt, pt, n_steps)
+            p1, u1 = pu[1][0], pu[1][1]
+            p0, u0 = bs_pt[0], bs_pt[1]
+            w = par_trans(u1, p1, None, p0) - u0
+            v = metric.log(point=p1, base_point=p0)
+            return n_steps * gs.array([v, w])
+
+        n_jobs = min(os.cpu_count(), len(pts))
+        with Parallel(n_jobs=n_jobs, verbose=0) as parallel:
+            rslt = parallel(_log(pt, bs_pts[i % len(bs_pts)])
+                            for i, pt in enumerate(pts))
+
+        return gs.reshape(gs.array(rslt), point.shape)
+
+    def geodesic_discrete(self, initial_point, end_point, n_steps=N_STEPS, **kwargs):
```

Comment:
To bring the parallelization here, we can simply make this method private and create a public one with the same name that handles  the parallelization and calls this method.